### PR TITLE
Add some type annotations and fix lints

### PR DESCRIPTION
**Pull request by Ericson2314**
The big thing is only `PerMachineDefaultable` needs to be optional across the board, not `PerMachine`. \[I later want to make it so that default_missing and friends produce a new object, so we don't store any `PerMachineDefaultable`s, just `PerMachines`. In other words, `PerMachineDefaultable` would be a factory of sorts.]

CC @dcbaker

### CODE REVIEW

File: mesonbuild/build.py [hunk 1]
```diff
@@ -450,7 +450,7 @@ def __init__(self, name, subdir, subproject, is_cross, sources, objects, environ
         self.is_unity = unity_opt == 'on' or (unity_opt == 'subprojects' and subproject != '')
         self.environment = environment
         self.sources = []
-        self.compilers = OrderedDict()
+        self.compilers = OrderedDict() # type: OrderedDict[str, Compiler]
```
Comment (dcbaker) [inline mesonbuild/build.py]: this needs to be typing.OrderedDict or mypy will complain.

Comment (Ericson2314) [reply mesonbuild/build.py]: Actually it didn't:
```
mesonbuild/build.py:452: error: Need type annotation for 'sources'
mesonbuild/build.py:454: error: Need type annotation for 'objects'
```

Comment (Ericson2314) [reply mesonbuild/build.py]: `mypy --strict --follow-imports silent --python-version 3.5 mesonbuild/build.py`. Tried again with `--python-version 3.5` this time, and 3.5-built mypy, and still no problem with line 453 just like above.

**Comment by Ericson2314** [comment]
`mypy --strict --follow-imports silent --python-version 3.5 mesonbuild/{envconfig.py,mlog.py}` indeed show no output.

**Pull request merged**